### ISSUE TITLE
lcov-1.15 version got stuck when collecting coverage and consumed 100% CPU

### ISSUE THREAD

**Issue by qiuhaonan** [linux-test-project/lcov#131]
![image](https://user-images.githubusercontent.com/20633488/136881387-ab73270c-d02a-411a-9d3b-4d4e4b996b73.png)
I updated lcov from 1.13 to 1.15 and run coverage collection task and used --rc geninfo_no_exception_branch=1 to filter c++ exception branch.
It ran over 12 hours and seemd to get stuck somewhere.
I searched some working experience that advised "--rc geninfo_gcov_all_blocks=0".
Any idea?

**Comment by qiuhaonan** [comment]
The main problems came from template class / functions.

1. The construction function of template class in our code could generate 500+ branches...., this class simply inherited a pure virtual class and had 6 template parameters (uint32_t or pointer type) and 6 members (uint32_t, mutex, slotvec)

2. A very simple conditional judgement in a template functions ("if (a == b)", a and b were uint32_t type) could also generate 100+ branches...

3. End of namespace or deconstruction function also generated 4 or more branches...

Though they were "reasonable" in low-level aspects, we could not selectively "disable or filter" them now. For every instance of template class or functions, it was hard to guarantee high branch coverage...

Basically, we expected to check if we executed simple judgements such as "if...else, switch, do...while, while, for or foreach". And for template, are there possible methods to bypass checking them? 

Currently, we could only skip files which only contain template class /functions, or write many "EXCL_BR_BEGIN" and "EXCL_BR_END" to help bypass checking them.

**Comment by henry2cox** [comment]
Interesting.  I think that we too have seen part of the problem that you describe – but we do not see some of the others.

Specifically:

-          “if (a == b) …”

o   will generate just two branches when ‘a’ and ‘b’ are integer types.
I’m curious why you are seeing very different branch counts here.  Does this happen everywhere?  Which compiler and binutils version? Are you using some unusual compile flags?
We _do_ see such counts in Verilog code, when we tell the simulator to blow up the vectors – but I don’t think you are looking at Verilog.

o   If ‘a’ and/or ‘b’ are classes which do fun things with operator==, conversion operators, etc – then there may be a lot of additional branches generated by the compiler – for example, exception return handling.
These we handle by filtering.

o   If a and b are template parameters – then it is quite likely that the compiler will evaluate at compile time (and there will be no branch in the code).

-          I think that the “construction of template function” case you mention is also related to implicit branches generated by the compiler (..which we handle similarly, via filtering).

-          I think that “end of namespace or deconstruction function…” meant “destructor calls at close of scope”

o   I think the branches here are implicitly generated by the compiler and are also related to exception handling.  We remove them by filtering.

The issue with the implicit branches is that they are quite often unreachable even if you were willing to devote heroic effort to try to test them.
Most of my users care only about the branches they explicitly wrote - and more-or-less believe that the compiler will generate correct code for exception handling, destructor calls, etc.  We do want to test exception handlers as well – but that is a slightly different problem, I think.

Given the above:  I had implemented a hack into my fork of lcov (…not yet merged back to master, unfortunately).
The hack simply looks at the source code and removes all the “BRDA” records that correspond to lines which do not contain conditionals.
For example:  neither the closing brace of a scope nor a straightforward function call statement contains a conditional – so branch records on those lines will be dropped.
The hack is pretty simple – and is easily defeated by a bloody-minded user or by complicated C++ code.  In those cases – we either live with the sub-optimal result, insert ‘exclude’ directives, restructure the code slightly to make the filter function happy – or enhance the filter implementation to handle more cases.

One can certainly argue that the filter approach is hacky.  Unfortunately, lcov/branch coverage is not actionable for C++ code without it.

I hope this helps

Henry

**Comment by qiuhaonan** [comment]
Thank you very much. I will try it.

I felt interesting to develop the lcov tool :)
There were many optimizations in code itself. It was a long-term work to do for large project.

Thanks.

Haonan
